File: fbiuniversal.py
K = int(input())

for k in range(K):
    line = input()
    inp = line.split()
    dset = inp[1]

    digits = list("0123456789ACDEFHJKLMNPRTVWX")
    dtoi = {};
    for i in range(len(digits)):
        dtoi[digits[i]] = i

    num = dset[0:8]
    check = dset[8]

    c = 2 * dtoi[num[0]] + \
        4 * dtoi[num[1]] + \
        5 * dtoi[num[2]] + \
        7 * dtoi[num[3]] + \
        8 * dtoi[num[4]] + \
        10 * dtoi[num[5]] + \
        11 * dtoi[num[6]] + \
        13 * dtoi[num[7]]  
    c = c % 27

    if c != dtoi[check]:
        print(inp[0], "Invalid")
        continue

    val = 0
    for i in range (len(num)):
        val += 27 ** (len(num) - 1 - i) * dtoi[num[i]]

    print(inp[0], val)

    

File: fbiuniversal.cpp
#include <iostream>
#include <cmath>
#include <unordered_map>
#include <functional>

using namespace std;

int main()
{
    string digs = "0123456789ACDEFHJKLMNPRTVWX";
    unordered_map<char, int> val;

    for(int i = 0 ; i < digs.size() ; ++i)
    {
        val[digs[i]] = i;
    }

    auto check = [&](string num) {
        return ((   2 * val[num[0]] + 4 * val[num[1]] + 5 * val[num[2]] + 
                    7 * val[num[3]] + 8 * val[num[4]] + 10 * val[num[5]] + 
                    11 * val[num[6]] + 13 * val[num[7]] ) % 27 );
    };

    int t; cin >> t;
    while (t--)
    {
        int k; cin >> k; cout << k << " ";

        string dat; cin >> dat;

        string num = dat.substr(0,8);
        int c = check(num);

        if (val[dat[8]] != c) { cout << "Invalid\n"; continue; }
        
        long long v = 0;
        for (int i = 0 ; i < num.size(); ++i)
        {
            v += (powl(27, i) * val[num[num.size() - 1 - i]]);
        }

        cout << v << '\n';
    }
}
